File: code-runner/src/sdks/runner-sdk.cs
// https://github.com/ultraslayyy/I-am-bored/tree/info/code-runner/src/sdks/runner-sdk.cs
// Please don't delete the above line, to credit me. But I can't stop you.
using System.Diagnostics;
using System.Net.Sockets;
using System.Text;
using System.Text.Json;

namespace RunnerClient;

public enum Lang
{
    Python,
    Python3,
    Javascript,
    Typescript,
    Cpp,
    C,
    Java,
    CSharp,
    Go,
    Kotlin,
    Swift,
    Rust,
    Ruby,
    Php,
    Dart,
    Scala,
    Elixir,
    Erlang,
    Racket,
    Bash,
    MySql,
    MsSql,
    Postgres,
    Oracle,
    Pandas
}

public class RunResult
{
    public string Stdout { get; set; }
    public string Stderr { get; set; }
    public int? ExitCode { get; set; }
    public bool Success { get; set; }
    public double TimeMs { get; set; }
    public long MemoryKb { get; set; }
}

public class TestCase
{
    public string Input { get; set; }
    public string Expected { get; set; }
}

public class TestCaseOptions
{
    public bool Parallel { get; set; } = false;
    public int ConcurrencyLimit { get; set; } = 10;
    public TestCase[] Cases { get; set; }
}

public class TestCaseResult : RunResult
{
    public string Input { get; set; }
    public string Expected { get; set; }
    public bool Passed { get; set; }
}

public class RunnerClient(string host = "127.0.0.1", int port = 4000)
{
    private readonly string _host = host;
    private readonly int _port = port;
    private Process serverProcess;

    public void StartServer(string filePath = "runner-server.ts", bool createWindow = false)
    {
        ProcessStartInfo startInfo = new()
        {
            FileName = "node",
            Arguments = filePath,
            RedirectStandardOutput = true,
            RedirectStandardError = true,
            UseShellExecute = false,
            CreateNoWindow = !createWindow
        };

        serverProcess = new()
        { StartInfo = startInfo };
        process.OutputDataReceived += (sender, e) => { if (e.Data != null) Console.WriteLine(e.Data); };
        process.ErrorDataReceived += (sender, e) => { if (e.Data != null) Console.Error.WriteLine(e.Data); };

        process.Start();
        process.BeginOutputReadLine();
        process.BeginErrorReadLine();

        Console.WriteLine($"Started server with PID {serverProcess.Id}");
    }

    public void StopServer()
    {
        if (serverProcess != null && !serverProcess.HasExited)
        {
            serverProcess.Kill();
            serverProcess.WaitForExit();
            Console.WriteLine("Server stopped.");
        }
        else
        {
            Console.WriteLine("Server is not running.");
        }
    }

    public async Task<RunResult> RunCodeLocalAsync(string code, Lang language, string input = "")
    {
        return await SendRequestAsync<RunResult>(code, language, input, null, false);
    }

    public async Task<TestCaseResult[]> RunCodeLocalAsync(string code, Lang language, TestCaseOptions testCases)
    {
        return await SendRequestAsync<TestCaseResult[]>(code, language, "", testCases, false);
    }

    public async Task<RunResult> RunCodeDockerAsync(string code, Lang language, string input = "")
    {
        return await SendRequestAsync<RunResult>(code, language, input, null, true);
    }

    public async Task<TestCaseResult[]> RunCodeDockerAsync(string code, Lang language, TestCaseOptions testCases)
    {
        return await SendRequestAsync<TestCaseResult[]>(code, language, "", testCases, true);
    }

    private async Task<T> SendRequestAsync<T>(string code, Lang language, string input, object testCases, bool useDocker)
    {
        var request = new
        {
            code,
            language = language.ToString().ToLower(),
            input,
            testCases,
            runner = useDocker ? "docker" : "local"
        };

        string jsonRequest = JsonSerializer.Serialize(request) + "\n";

        using TcpClient client = new();
        await client.ConnectAsync(_host, _port);
        using NetworkStream stream = client.GetStream();

        byte[] requestBytes = Encoding.UTF8.GetBytes(jsonRequest);
        await stream.WriteAsync(requestBytes);

        StringBuilder responseBuilder = new();
        byte[] buffer = new byte[1024];
        int bytesRead;
        do
        {
            bytesRead = await stream.ReadAsync(buffer, 0, buffer.Length);
            responseBuilder.Append(Encoding.UTF8.GetString(buffer, 0, bytesRead));
        } while (!responseBuilder.ToString().Contains('\n'));

        string jsonResponse = responseBuilder.ToString().Trim();
        return JsonSerializer.Deserialize<T>(jsonResponse);
    }
}

File: code-runner/src/sdks/runner-sdk.py
# https://github.com/ultraslayyy/I-am-bored/tree/info/code-runner/src/sdks/runner-sdk.py
# Please don't delete the above line, to credit me. But I can't stop you.
import socket, json, subprocess, sys, threading
from typing import Optional, List, Dict, Any, Union

TestCase = Dict[str, Any]
TestCaseOptions = Dict[str, Any]
RunResult = Dict[str, Any]

class RunnerClient:
    def __init__(self, host: str = 'localhost', port: int = 4000):
        self.host = host
        self.port = port
        self._server_process: subprocess.Popen | None = None

    def start_server(self, file_path: str = "runner-server.js"):
        if self._server_process is not None and self._server_process.poll() is None:
            print("Server is already running.")
            return
        
        self._server_process = subprocess.Popen(
            ["node", file_path],
            stdout=subprocess.PIPE,
            stderr=subprocess.PIPE,
            text=True,
            shell=False
        )

        def _print_stream(stream, is_err=False):
            for line in stream:
                if line:
                    print(line, end='', file=sys.stderr if is_err else sys.stdout)
        
        threading.Thread(target=_print_stream, args=(self._server_process.stdout,), daemon=True).start()
        threading.Thread(target=_print_stream, args=(self._server_process.stderr, True), daemon=True).start()

        print(f"Started server with PID {self._server_process.pid}")
    
    def stop_server(self):
        if self._server_process is not None and self._server_process.poll() is None:
            self._server_process.terminate()
            self._server_process.wait()
            print("Server stopped.")
        self._server_process = None

    def _send_request(self, payload: Dict[str, Any]) -> Union[RunResult, List[RunResult]]:
        with socket.create_connection((self.host, self.port)) as sock:
            sock.sendall((json.dumps(payload) + '\n').encode())
            data = b''
            while True:
                chunk = sock.recv(4096)
                if not chunk:
                    break
                data += chunk
                if b'\n' in data:
                    break
            response = json.loads(data.decode().strip())
            if not response.get("success", False):
                raise RuntimeError(response.get("error", "Unknown error"))
            return response["result"]

    def run_code_local(self, code: str, language: str, filenamePrefix: str, input: Optional[str] = '', test_cases: Optional[TestCaseOptions] = None) -> Union[RunResult, List[RunResult]]:
        payload = {
            "runner": "local",
            "code": code,
            "language": language
        }

        if test_cases:
            payload["testCases"] = test_cases
        else:
            payload["input"] = input
        
        if filenamePrefix:
            payload["filenamePrefix"] = filenamePrefix

        return self._send_request(payload)

    def run_code_docker(self, code: str, language: str, filenamePrefix: str, input: Optional[str] = '', test_cases: Optional[TestCaseOptions] = None) -> Union[RunResult, List[RunResult]]:
        payload = {
            "runner": "docker",
            "code": code,
            "language": language
        }

        if test_cases:
            payload["testCases"] = test_cases
        else:
            payload["input"] = input

        if filenamePrefix:
            payload["filenamePrefix"] = filenamePrefix

        return self._send_request(payload)